src/Google.Api.Gax/Api/Gax/PathTemplate.cs

```diff
@@ -0,0 +1,440 @@
+                        if (nameSegment != segment.Value)
+                        {
+                            result = null;
+                            return $"Name does not match template in literal segment: '{nameSegment}' != '{segment.Value}'";
+                        }
+                        break;
+                    case SegmentKind.Wildcard:
+                        // Could use segment.ValidateWildcard, but the exception wouldn't be as clean.
+                        var value = nameSegments[nameSegmentIndex++];
+                        if (value == "")
+                        {
+                            result = null;
+                            return "Name does not match template: wildcard segment is empty";
+                        }
+                        resourceIds[resourceIdIndex++] = value;
+                        break;
+                    case SegmentKind.PathWildcard:
+                        // Work out how many segments to consume based on the number of segments in the template and the
+                        // actual number of segments in the specified name
+                        int count = nameSegments.Length - _segments.Count + 1;
+                        // Make the common case more efficient
+                        if (count == 1)
+                        {
+                            resourceIds[resourceIdIndex++] = nameSegments[nameSegmentIndex++];
+                        }
+                        else
+                        {
+                            resourceIds[resourceIdIndex++] = string.Join("/", nameSegments.Skip(nameSegmentIndex).Take(count));
+                            nameSegmentIndex += count;
+                        }
+                        break;
+                }
+            }
+            result = ResourceName.CreateWithShallowCopy(this, serviceName, resourceIds);
+            return null; // Success!
+        }
+
+        /// <summary>
+        /// Returns the textual representation of this template.
+        /// </summary>
+        /// <returns>The same textual representation that this template was initially constructed with.</returns>
+        public override string ToString() => _originalTemplate;
+
+        private enum SegmentKind
+        {
+            /// <summary>
+            /// A literal path segment.
+            /// </summary>
+            Literal,
+
+            /// <summary>
+            /// A simple wildcard ('*').
+            /// </summary>
+            Wildcard,
+
+            /// <summary>
+            /// A path wildcard ('**').
+            /// </summary>
+            PathWildcard,
+        }
+
+        /// <summary>
+        /// A segment of a path.
+        /// </summary>
+        private sealed class Segment
+        {
+            private static readonly Segment UnnamedWildcard = new Segment(SegmentKind.Wildcard, null);
+            private static readonly Segment UnnamedPathWildcard = new Segment(SegmentKind.PathWildcard, null);
```

Comment:
Done. I think I originally had these as internal properties, hence the different names...